src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+// Main method
+//
+
+    public void analyzeTree(Env<AttrContext> env) {
+        this.analyzeTree(env, env.tree);
+    }
+
+    public void analyzeTree(Env<AttrContext> env, JCTree tree) {
+
+        // Sanity check
+        Assert.check(this.checkInvariants(false, false));
+
+        // Short circuit if warnings are totally disabled
+        if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+            return;
+
+        // Build a mapping from symbols of methods to their declarations.
+        // Classify all ctors and methods as analyzable and/or invokable.
+        // Track which methods and variables have warnings are suppressed.
+        new TreeScanner() {
+
+            private Lint lint = ThisEscapeAnalyzer.this.lint;
+            private JCClassDecl currentClass;
+            private boolean privateOuter;
+
+            @Override
+            public void visitClassDef(JCClassDecl tree) {
+                final JCClassDecl currentClassPrev = this.currentClass;
+                final boolean privateOuterPrev = this.privateOuter;
+                final Lint lintPrev = this.lint;
+                this.lint = this.lint.augment(tree.sym);
+                try {
+                    this.currentClass = tree;
+                    this.privateOuter |= tree.sym.isAnonymous();
+                    this.privateOuter |= (tree.mods.flags & Flags.PRIVATE) != 0;
+
+                    // Recurse
+                    super.visitClassDef(tree);
+                } finally {
+                    this.currentClass = currentClassPrev;
+                    this.privateOuter = privateOuterPrev;
+                    this.lint = lintPrev;
+                }
+            }
+
+            @Override
+            public void visitVarDef(JCVariableDecl tree) {
+                final Lint lintPrev = this.lint;
+                this.lint = this.lint.augment(tree.sym);
+                try {
+                    // Track warning suppression
+                    if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+                        ThisEscapeAnalyzer.this.suppressed.add(tree.sym);
+
+                    // Recurse
+                    super.visitVarDef(tree);
+                } finally {
+                    this.lint = lintPrev;
+                }
+            }
+
+            @Override
+            public void visitMethodDef(JCMethodDecl tree) {
+                final Lint lintPrev = this.lint;
+                this.lint = this.lint.augment(tree.sym);
+                try {
+                    // Track warning suppression
+                    if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+                        ThisEscapeAnalyzer.this.suppressed.add(tree.sym);
+
+                    // Determine if this is a constructor we should analyze
+                    final boolean analyzable = this.currentClassIsExternallyExtendable() &&
+                        TreeInfo.isConstructor(tree) &&
+                        !tree.sym.isPrivate() &&
```

Comment:
> If we have a class with a private constructor and public static factory invoking said constructor, and the constructor makes this escape, isn't that an issue we should detect?

A static factory method will not create a subclassed instance, so there's no 'this' escape problem.

But I admit I completely missed factory methods as a potential thing to worry about.

Is it possible for a leak to be missed due to the use of a factory method?

Hmm. I can't immediately think of how, but if you can come up with an example please share.

Reply:
I guess what I'm thinking about:
```
class Foo {
    private Foo() {
        m(this);
    }
    
    public void m() { ... } // overridable
    
    static Foo makeFoo() { return new Foo(); }
}
```

Reply:
> I guess what I'm thinking about:

No leak is possible in that example.
* `new Foo()` creates an instance of `Foo` (not a subclass of `Foo`) therefore `m()` is not overridden
* Any subclass of `Foo` that may exist in the outside world cannot use the `Foo()` constructor that leaks because it's private

Reply:
but what if `m` is a static method in a separate compilation unit?

Reply:
Caring about the proper initialization of any class in the _current_ compilation unit is an explicit non-goal.

We only care about bugs where a superclass and subclass are in separate compilation units.

Reply:
So, to clarify, in this case:

```java
import java.util.*;

class B {
    final Object ref;

     private B(Object ref) {
          Foo.consume(this);
          this.ref = ref;
      }
 }
```

Even though `this` leaks to a method clearly before it is fully initialized, we do not care because there can be no subclass involved observing this. I guess I was confused because, while subclasses are a particularly sneaky case where uninitialized values can show up, the above leak seems potentially dangerous as well - we're effectively leaking a class whose final field has not been initialized!

That said, if that was discussed, and it was decided for the warning not to deal with this case, that's ok.

Reply:
Perhaps my confusion might come from the name `this-escape` of the lint warning - which seems overpromising in this respect. But I looked at the description of the lint warning using `javac --help-lint` and I got this:

```
    this-escape          Warn when a constructor invokes a method that could be overriden in a subclass;
```

Which indeed aligns well with what this PR is doing. So that's ok.

One minor nit - all the other Lint descriptions end with a period, but this one end with semi-colon:

```
...
static               Warn about accessing a static member using an instance.
    strictfp             Warn about unnecessary use of the strictfp modifier.
    synchronization      Warn about synchronization attempts on instances of value-based classes.
    text-blocks          Warn about inconsistent white space characters in text block indentation.
    this-escape          Warn when a constructor invokes a method that could be overriden in a subclass;
such a method would execute before the subclass constructor completes its initialization.
    try                  Warn about issues relating to use of try blocks (i.e. try-with-resources).
...
```

You might want to fix that.

Reply:
Something seems to be up with the lint description for this-escape - compare this:

```
  serial               Warn about Serializable classes that do not have a serialVersionUID field. 
                         Also warn about other suspect declarations in Serializable and Externalizable classes and interfaces.
```
with this:

```
  this-escape          Warn when a constructor invokes a method that could be overriden in a subclass;
such a method would execute before the subclass constructor completes its initialization.
```

Indentation seems to be missing, which causes readability issues in the `--help-lint` output.

Reply:
> I guess I was confused because, while subclasses are a particularly sneaky case where uninitialized values can show up, the above leak seems potentially dangerous as well...

Yes - and this very question did come up in the discussions around this warning (see amber-dev).

The decision was to go with drawing the "warning boundary" at the compilation unit. The reasoning is that (a) this aligns with the compiler's "knowledge boundary", i.e., we can know for sure from code inspection, and also (b) focuses the warning on the particularly pernicious aspect of these bugs, which is that they arise from the non-obvious interaction among two or more files - even when looking at any single one of those files, there doesn't seem to be any apparent problem. In other words, we decided "not to try to save any single source code from itself".

But I think it's still an interesting question. Maybe experience will provide more guidance over time.

Reply:
> Something seems to be up with the lint description for this-escape - compare this:

Oops, will fix - thanks.

Reply:
> The decision was to go with drawing the "warning boundary" at the compilation unit. The reasoning is that (a) this aligns with the compiler's "knowledge boundary", i.e., we can know for sure from code inspection, and also (b) focuses the warning on the particularly pernicious aspect of these bugs, which is that they arise from the non-obvious interaction among two or more files

Sorry for being picky - you mention this "compilation unit" boundary before, but this is not really the reason here. Note that in my example B constructor calls out to a static method that is "outside" the boundary. The reason as to why my example is not flagged is simply that "escaping" is defined as "escaping into a subclass method", not just "escaping from the constructor (into some other compilation unit)".

Reply:
Oops, you're right, I answered the wrong question so to speak.

The "must involve a subclass" requirement is another dimension on which a "boundary" was declared. This was also part of the original discussion.

So yes the requirement is: "requires involvement of a subclass" **AND** "that subclass lives in a separate compilation unit".